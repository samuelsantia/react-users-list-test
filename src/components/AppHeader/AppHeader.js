import React from 'react';
import styled, { css } from 'styled-components/macro';
import PropTypes from 'prop-types';
import { medias } from '../../themes/utils';

/**
 * sets display flex
 * @param {string} justify justify-content prop
 */
const basicHeaderFlex = (justify) => css`
  display: flex;
  align-items: center;
  justify-content: ${justify};
`;

/** header container */
export const HeaderWrapper = styled.header`
  ${basicHeaderFlex('space-between')};
  padding: var(--app-header-padding, 1rem);
  background-color: var(--app-header-background-color, white);
  box-shadow: var(--app-header-shadow, none);
`;

/** header left side container */
export const HeaderStartSide = styled.div`
  ${basicHeaderFlex('flex-start')};
`;

/** header center side container */
export const HeaderCenterSide = styled.div`
  display: none;

  @media ${medias.tablet} {
    ${basicHeaderFlex('center')};
    flex: 1;
  }
`;

/** header right side container */
export const HeaderEndSide = styled.div`
  ${basicHeaderFlex('flex-end')};
`;

/** Renders App main header */
function AppHeader({ start, children, end }) {
  return (
    <HeaderWrapper>
      <HeaderStartSide>{start}</HeaderStartSide>
      <HeaderCenterSide>{children}</HeaderCenterSide>
      <HeaderEndSide>{end}</HeaderEndSide>
    </HeaderWrapper>
  );
}

AppHeader.propTypes = {
  /** left content */
  start: PropTypes.node,
  /** center content */
  children: PropTypes.node,
  /** right content */
  end: PropTypes.node,
};

export default AppHeader;
