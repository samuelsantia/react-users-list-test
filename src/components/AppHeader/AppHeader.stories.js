import React from 'react';

import Typography from '../Typography';
import AppHeader from './AppHeader';

export default {
  title: 'AppHeader',
  component: AppHeader,
};

export const basic = () => (
  <AppHeader
    start={
      <Typography variant="h2">
        <Typography variant="h2" tag="strong" weight="bold">
          App
        </Typography>
        Logo
      </Typography>
    }
    end={
      <Typography variant="h6" tag="p">
        pedro@email.com
      </Typography>
    }
  >
    <Typography variant="h5">Center Content</Typography>
  </AppHeader>
);
