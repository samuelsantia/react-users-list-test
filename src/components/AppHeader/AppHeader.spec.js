import React from 'react';
import { render } from '@testing-library/react';
import { find } from 'styled-components/test-utils';
import 'jest-styled-components';

import AppHeader, {
  HeaderWrapper,
  HeaderStartSide,
  HeaderCenterSide,
  HeaderEndSide,
} from './AppHeader';

const getTag = (element) => element.tagName.toLowerCase();
const findStyled = (Styled) => (element) => find(element, Styled);
const findWrapper = findStyled(HeaderWrapper);
const findStartSide = findStyled(HeaderStartSide);
const findCenterSide = findStyled(HeaderCenterSide);
const findEndSide = findStyled(HeaderEndSide);

describe('<AppHeader />', () => {
  it('should render styled wrapper with sides', () => {
    const { container } = render(<AppHeader />);

    const wrapper = findWrapper(container);
    const startSide = findStartSide(wrapper);
    const centerSide = findCenterSide(wrapper);
    const endSide = findEndSide(wrapper);

    expect(getTag(wrapper)).toBe('header');
    // right order
    expect([...wrapper.children]).toEqual([startSide, centerSide, endSide]);
  });

  it('should render start content', () => {
    const content = <p>test</p>;
    const { container } = render(<AppHeader start={content} />);

    const rendererContent = findStartSide(container).firstChild;
    expect(getTag(rendererContent)).toBe('p');
    expect(rendererContent).toHaveTextContent('test');
  });

  it('should render center content passed as children', () => {
    const content = <p>test</p>;
    const { container } = render(<AppHeader>{content}</AppHeader>);

    const rendererContent = findCenterSide(container).firstChild;
    expect(getTag(rendererContent)).toBe('p');
    expect(rendererContent).toHaveTextContent('test');
  });

  it('should render end content', () => {
    const content = <p>test</p>;
    const { container } = render(<AppHeader end={content} />);

    const rendererContent = findEndSide(container).firstChild;
    expect(getTag(rendererContent)).toBe('p');
    expect(rendererContent).toHaveTextContent('test');
  });

  it('should match snapshot', () => {
    const { container } = render(
      <AppHeader start="Start" end="Ends">
        Content
      </AppHeader>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
