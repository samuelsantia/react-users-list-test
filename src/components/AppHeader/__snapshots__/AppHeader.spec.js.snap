// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppHeader /> should match snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: var(--app-header-padding,1rem);
  background-color: var(--app-header-background-color,white);
  box-shadow: var(--app-header-shadow,none);
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

<header
  class="c0"
>
  <div
    class="c1"
  >
    Start
  </div>
  <div
    class="c2"
  >
    Content
  </div>
  <div
    class="c3"
  >
    Ends
  </div>
</header>
`;
